Update ScrollPanel to extend AbstractWidget

Caltinor opened the pull request:
This PR updates the NF class to incorporate the Layout system.   

Key changes:
- un-finalized dimension values
- implemented methods from new abstract class (mainly `LayoutElement` methods)
- refactored `left` and `top` as `x` and `y` respectively to match terminology used elsewhere
- removed `bottom` and `right` variables in favor of methods to calculate them from dimension variables dynamically
- Adds new `Component` constructor parameter to match widget pattern
- ScrollPanel no longer extends `AbstractContainerEventHandler` but only `ContainerEventHandler`.  implementation from ACEH was cloned into ScrollPanel for functionality parity with the former abstract class.

Caltinor committed:
implement LayoutElement

this also includes a refactor of the `left` and `top` variables to better match the x/y term usage in other areas of the codebase

Caltinor committed:
fix implementing classes

Caltinor committed:
formatting go brrrrr

Caltinor committed:
update to extend AbstractWidget

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/ModListScreen.java:
```diff
@@ -269,7 +269,7 @@ public void init() {
         this.modList = new ModListWidget(this, listWidth, fullButtonHeight, search.getY() - getFontRenderer().lineHeight - PADDING);
         this.modList.setX(6);
-        this.modInfo = new InfoPanel(this.minecraft, modInfoWidth, this.height - PADDING - fullButtonHeight, PADDING);
+        this.modInfo = new InfoPanel(this.minecraft, modInfoWidth, this.height - PADDING - fullButtonHeight, PADDING, Component.literal("mod list scrolling panel"));
```
Can't we just supply an empty component here since it's not being used or rendered?

Caltinor replied:
AbstractWidget uses the message as a default narration element.  I figured this was good for accessibility.

ChampionAsh5357 replied:
It depends whether the panel itself is a highlightable element. Take, for example, `AbstractSelectionList`, which is a similar sort of list selector. The widget supplies an empty component for itself while it children are handled within the `updateWidgetNarration` of its subtype. In this particular instance, the panel is basically reading out its contents of the mod information, so for proper support, it would need to read the lines in its subimplementations.

So, there are two methods. Either the panel reads out the narration when it's info is updated, or it shouldn't do anything at all.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/ModMismatchDisconnectedScreen.java:
```diff
@@ -80,7 +80,7 @@ protected void init() {
         int upperButtonHeight = Math.min((this.height + this.listHeight) / 2 + 25, this.height - 50);
         int lowerButtonHeight = Math.min((this.height + this.listHeight) / 2 + 50, this.height - 25);
-        this.addRenderableWidget(this.scrollList = new MismatchInfoPanel(minecraft, listWidth, listHeight, (this.height - this.listHeight) / 2, listLeft));
+        this.addRenderableWidget(this.scrollList = new MismatchInfoPanel(minecraft, listWidth, listHeight, (this.height - this.listHeight) / 2, listLeft, Component.literal("mismatched info scroll panel")));
```
Can't we just supply an empty component here since it's not being used or rendered?

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -7,26 +7,26 @@
+    protected int width;
+    protected int height;
+    protected int y;
+    protected int x;
```
What are these being double declared if we can just use the ones in the `AbstractWidget`?

Caltinor replied:
absolutely.  i started this PR by just implementing `LayoutElement` and later replaced the abstract class with `AbstractWidget` so i overlooked the duplication.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -42,36 +42,36 @@ public abstract class ScrollPanel extends AbstractContainerEventHandler implemen
      */
-    public ScrollPanel(Minecraft client, int width, int height, int top, int left) {
-        this(client, width, height, top, left, 4);
+    public ScrollPanel(Minecraft client, int width, int height, int y, int left, Component component) {
```
Documentation for the component parameter?

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -247,4 +251,81 @@ protected void drawGradientRect(GuiGraphics guiGraphics, int left, int top, int
     }
+
+    @Override
+    public void setX(int p_265236_) {
```
Only gonna mention this once, but name the parameters appropriately.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -247,4 +251,81 @@ protected void drawGradientRect(GuiGraphics guiGraphics, int left, int top, int
+
+    @Override
+    public NarrationPriority narrationPriority() {
+        return null;
```
I'm pretty sure this can throw an error.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -7,26 +7,26 @@
  * Abstract scroll panel class.
  */
-public abstract class ScrollPanel extends AbstractContainerEventHandler implements Renderable, NarratableEntry {
+public abstract class ScrollPanel extends AbstractWidget implements ContainerEventHandler {
```
Does this even need to implement `ContainerEventHandler` when it already implements `GuiEventListener`? I don't think it makes use of any of the extensions.

Caltinor replied:
`ContainerEventHandler` has a lot of default behavior and some extra methods that `GuiEventHandler` does not have.  I did not evaluate whether those default implementations were critical to this class.  I mainly kept it for compatibility since anything that is looking for an implementation of `ContainerEventHandler` would not implicitly accept its superclass.

ChampionAsh5357 replied:
Then why not use `AbstractContainerWidget` as the base instead of `AbstractWidget`? That implements the container event handler while also defining a scroll area.

Caltinor replied:
Honestly, didn't realize that existed.  implementation updated.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -247,4 +251,81 @@ protected void drawGradientRect(GuiGraphics guiGraphics, int left, int top, int
+    @Override
+    public int getHeight() {
+        return height;
+    }
```
Why?

Caltinor replied:
see https://github.com/neoforged/NeoForge/pull/2734#discussion_r2442573758

Caltinor committed:
cleanup from review

Caltinor committed:
Merge branch '1.21.x' into 1.21.x

Caltinor committed:
fix formatting

Caltinor committed:
further simplify

replacing basic AbstractWidget implementation with AbstractContainerWidget instead

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -212,24 +211,24 @@ public boolean mouseDragged(MouseButtonEvent event, double deltaX, double deltaY
 
-            int barTop = (int) this.scrollDistance * (height - barHeight) / extraHeight + this.top;
-            if (barTop < this.top) {
-                barTop = this.top;
+            int barTop = (int) this.scrollDistance * (height - barHeight) / extraHeight + this.getY();
+            if (barTop < this.getY()) {
+                barTop = this.getY();
             }
 
-            guiGraphics.fill(this.barLeft, this.top, this.barLeft + this.barWidth, this.bottom, this.barBgColor);
+            guiGraphics.fill(this.barLeft, this.getY(), this.barLeft + this.barWidth, this.getBottom(), this.barBgColor);
```
Since this is responsible for rendering the scrollbar, it probably makes more sense to override and replace `renderScrollbar` with it and call here.

ChampionAsh5357 commented on src/client/java/net/neoforged/neoforge/client/gui/widget/ScrollPanel.java:
```diff
@@ -9,24 +9,18 @@
-    protected final int bottom;
-    protected final int right;
-    protected final int left;
     private boolean scrolling;
     protected float scrollDistance;
```
With the addition of the container widget, these fields are in its superclass. It should be updated to set and use those values instead.

Caltinor commented:
After realizing this is only used by NF and not adding value to users via the existence of `AbstractScrollWidget` i am closing this PR.
